### PR TITLE
RavenDB-22193 - Improve the performance of indexes with LoadDocument from different collections

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22193/Improve-the-performance-of-indexes-with-LoadDocument-from-different-collections

### Additional description

- When references from different collections are being changed, we might index the same document more than once, we now identify that and skip it (this is valid since we are in the same read transaction).
- Renewing a transaction will clear the already indexed documents (same as the previous behaviour).
- Check if we already indexed the document BEFORE we fetch it from the disk.
- Use `Slice` instead of materializing the item key to a `string`.

In the provided test we have `20K` documents with 2 references each (to the same 2 documents).
When those 2 are being changed, we would index all parent documents twice (`40K` map reference attemps), now we do it only once (`20K` map reference attemps).

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 1]
```diff
@@ -173,12 +180,11 @@ protected HandleReferencesBase(Index index, DocumentsStorage documentsStorage, I
...
 
+                            RenewTransactionIfNeeded(queryContext, batchContinuationResult, indexed, ref readTransaction);
+
                             var hasChanges = false;
                             earlyExit = false;
 
-                            var indexed = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-
-                            using (queryContext.OpenReadTransaction())
                             {
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: This can be removed, but doing so causes a leftward shift, giving the impression of more changes than actually exist.
Once this is reviewed, I'll format it correctly.

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 2]
```diff
@@ -137,6 +137,13 @@ protected HandleReferencesBase(Index index, DocumentsStorage documentsStorage, I
                 if (lastIndexedEtag == 0) // we haven't indexed yet, so we are skipping references for now
                     continue;
 
+                // When opening the read transaction for a single map covering all of its referenced collections,
+                // we can efficiently skip indexing documents that have already been indexed.
+                var readTransaction = queryContext.OpenReadTransaction();
+                var indexed = new HashSet<Slice>(SliceComparer.Instance);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs:143]: Previously we had `OrdinalIgnoreCase` while the `SliceComparer` doesn't have the notion of ignoring the casing. Do we care about that?

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 3]
```diff
@@ -354,6 +360,16 @@ bool CanContinueReferenceBatch()
                         _referencesState.Clear(earlyExit, actionType, collection, indexContext);
                     }
                 }
+                }
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: Formatting issue. I know that one you fix it you'll have ~200 lines changed but when you put it in a dedicated commit I think that will be fine

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 4]
```diff
@@ -385,22 +401,44 @@ private void UpdateReferences(TransactionOperationContext indexContext, string c
             }
         }
 
+        private static void RenewTransactionIfNeeded(QueryOperationContext queryContext, Index.CanContinueBatchResult batchContinuationResult, HashSet<Slice> indexed, ref IDisposable readTransaction)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs:402]: Why do you use `IDisposable` for the readTransaction instead of the actual transaction type

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 5]
```diff
@@ -385,22 +401,44 @@ private void UpdateReferences(TransactionOperationContext indexContext, string c
...
             var lastProcessedItemId = referenceState?.GetLastProcessedItemId(referencedItem);
             foreach (var key in _referencesStorage.GetItemKeysFromCollectionThatReference(collection, referencedItem.Key, indexContext.Transaction, lastProcessedItemId))
             {
-                var item = GetItem(queryContext.Documents, key);
-                if (item == null)
-                    continue;
-
-                if (indexed.Add(item.Id) == false)
+                // we check if we already indexed the document BEFORE we fetch it from the disk.
+                var clonedKey = key.Clone(queryContext.Documents.Allocator);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: Instead of always doing `Clone()` you could do the check with the usage of `Contains()`. If it's not there then you'll do `Clone()` and `indexed.Add()`

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs:143]: It doesn't matter. Those are the parent item IDs which are the same across different referenced collections.

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: done

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs:402]: I thought that `OpenReadTransaction()` returns `DocumentsTransaction` object but now I see it returns `IDisposable`

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: What do you think? Someone will come here one day and do the formatting anyway

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: done

**Review by arekpalinski — APPROVED** [review]

**Review by maciejaszyk — APPROVED** [review]

**Comment by arekpalinski** [comment]
@ppekrol I'm merging this but do the review

**Comment by redknightlois** [comment]
Performance metrics at: https://issues.hibernatingrhinos.com/issue/RDBQA-175/Check-RavenDB-22193-performance
